File: bench/04_optim.R
library(torch)

batch_size <- as.numeric(Sys.getenv("BATCH_SIZE", unset = "1000"))
device <- Sys.getenv("DEVICE", unset = "cpu")

w <- torch_randn(784, 512, requires_grad = TRUE, device=device)
x <- torch_randn(batch_size, 784, device=device)
opt <- optim_sgd(w, lr = 0.01)

f <- function() {
  for (i in 1:iter) {
    loss <- torch_sum(torch_mm(x, w))
    loss$backward()
    opt$step()
    opt$zero_grad()
  }

  invisible(NULL)
}

if (device == "cpu") {
  fn <- f
} else {
  fn <- function() {
    f()
    torch:::cuda_synchronize()
  }


iter <- 1
fn()
iter <- as.numeric(Sys.getenv("ITER", unset = "1000"))

cat(system.time(fn())[["elapsed"]])

File: bench/04_optim.py
import torch
import time
import os

batch_size = int(os.environ.get('BATCH_SIZE', "1000"))
device = os.environ.get("DEVICE", "cpu")

def mm ():
  x = torch.randn(batch_size, 784, device = device)
  w = torch.randn(784, 512, requires_grad = True, device = device)
  opt = torch.optim.SGD([w], lr = 0.01)
  for i in range(ir):
    loss = torch.sum(torch.mm(x, w))
    loss.backward()
    opt.step()
    opt.zero_grad()
    
ir = 1
mm()
ir = int(os.environ.get('ITER', "1000"))

start_time = time.time()
mm()
print((time.time() - start_time))

